src/core/server/rendering/create_metadata_factory.ts

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { take } from 'rxjs/operators';
+
+import { i18n } from '@kbn/i18n';
+
+import { mergeVariables } from './merge_variables';
+import { RenderingMetadata, RenderingProviderParams, PluginVariables } from './types';
+
+async function getVariables(
+  id: string,
+  pluginConfigs: PluginVariables[],
+  params: RenderingProviderParams
+) {
+  const config = (params.http.server as any).config();
+  const defaultVars =
+    params.legacyPlugins.uiExports.defaultInjectedVarProviders?.reduce(
+      (defaults, { fn, pluginSpec }) =>
+        mergeVariables(fn(params.http.server, pluginSpec.readConfigValue(config, [])), defaults),
+      {}
+    ) ?? {};
+  const injectedVars = mergeVariables(
+    defaultVars,
+    ...pluginConfigs,
+    await params.getVarsFor(id),
```

Comment:
This mechanism is untested, though see other comment about leaving this in legacy.